source/opt/loop_unroller.cpp

```diff
@@ -0,0 +1,553 @@
+
+  uint32_t phi_index = 5;
+  uint32_t phi_variable =
+      state_.previous_phi_->GetSingleWordOperand(phi_index - 1);
+  uint32_t phi_label = state_.previous_phi_->GetSingleWordOperand(phi_index);
+
+  ir::Instruction* original_phi = loop->GetInductionVariable();
+
+  // SetInOperands are offset by two.
+  // TODO: Work out why.
+  original_phi->SetInOperand(phi_index - 3, {phi_variable});
+  original_phi->SetInOperand(phi_index - 2, {phi_label});
+}
+
+void LoopUnrollerUtilsImpl::FullyUnroll(ir::Loop* loop) {
+  Unroll(loop, loop->NumIterations() - 1);
+
+  // When fully unrolling we can eliminate the last condition block.
+  FoldConditionBlock(state_.previous_condition_block_, 1);
+
+  // Delete the OpLoopMerge and remove the backedge to the header.
+  CloseUnrolledLoop(loop);
+
+  // Add the blocks to the function.
+  AddBlocksToFunction(loop->GetMergeBlock());
+
+  // Invalidate all analyses.
+  ir_context_->InvalidateAnalysesExceptFor(
+      ir::IRContext::Analysis::kAnalysisNone);
+}
+
+void LoopUnrollerUtilsImpl::CopyBasicBlock(ir::Loop* loop,
+                                           const ir::BasicBlock* itr) {
+  ir::BasicBlock* BB = itr->Clone(ir_context_);
+  // Assign each result a new unique ID and keep a mapping of the old ids to
+  // the new ones.
+  RemapResultIDs(loop, BB);
+
+  // If this is the continue block we are copying.
+  if (itr == loop->GetLatchBlock()) {
+    // Make the OpLoopMerge point to this block for the continue.
+    ir::Instruction* merge_inst = loop->GetHeaderBlock()->GetLoopMergeInst();
+    merge_inst->SetInOperand(1, {BB->id()});
+
+    state_.new_continue_block = BB;
+  }
+
+  // If this is the header block we are copying.
+  if (itr == loop->GetHeaderBlock()) {
+    state_.new_header_block = BB;
+
+    // Remove the loop merge instruction if it exists.
+    ir::Instruction* merge_inst = BB->GetLoopMergeInst();
+    if (merge_inst) ir_context_->KillInst(merge_inst);
+  }
+
+  // If this is the condition block we are copying.
+  if (itr == loop->GetConditionBlock()) {
+    state_.new_condition_block = BB;
+  }
+
+  // Add this block to the list of blocks to add to the function at the end of
+  // the unrolling process.
+  blocks_to_add_.push_back(std::unique_ptr<ir::BasicBlock>(BB));
+
+  // Keep tracking the old block via a map.
+  state_.new_blocks[itr->id()] = BB;
+}
+
+void LoopUnrollerUtilsImpl::CopyBody(ir::Loop* loop,
+                                     bool eliminate_conditions) {
+  const ir::Loop::BasicBlockOrderedListTy& basic_blocks =
+      loop->GetOrderedBlocks();
+
+  // Copy each basic block in the loop, give them new ids, and save state
+  // information.
+  for (const ir::BasicBlock* itr : basic_blocks) {
+    CopyBasicBlock(loop, itr);
+  }
+
+  // Set the previous continue block to point to the new header.
+  ir::Instruction& continue_branch = *state_.previous_continue_block_->tail();
+  continue_branch.SetInOperand(0, {state_.new_header_block->id()});
```

Comment:
That's also very specific and won't work after a rotate for instance.
you need to look for the back-edge

Reply:
Still needs to be updated.